File: camerax_android/lib/src/video/quality_selector_impl.dart
import 'package:camerax_android/src/api.dart';
import 'package:camerax_android/src/common.dart';
import 'package:camerax_android/src/core.dart';
import 'package:camerax_android/src/video.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

final class QualitySelectorImpl implements QualitySelector {
  final QualitySelectorProxyApi api;

  QualitySelectorImpl.internal(this.api);
}

final class QualitySelectorChannelImpl extends QualitySelectorChannel {
  @override
  Future<Size<int>?> getResolution(CameraInfo cameraInfo, Quality quality) =>
      QualitySelectorProxyApi.getResolution(
        cameraInfo.api,
        quality.api,
      ).then((e) => e?.impl);

  @override
  QualitySelector from(Quality quality, [FallbackStrategy? fallbackStrategy]) {
    final api = fallbackStrategy == null
        ? QualitySelectorProxyApi.from1(quality: quality.api)
        : QualitySelectorProxyApi.from2(
            quality: quality.api,
            fallbackStrategy: fallbackStrategy.api,
          );
    return QualitySelectorImpl.internal(api);
  }

  @override
  QualitySelector fromOrderedList(
    List<Quality> qualities, [
    FallbackStrategy? fallbackStrategy,
  ]) {
    final api = fallbackStrategy == null
        ? QualitySelectorProxyApi.fromOrderedList1(
            qualities: qualities.map((e) => e.api).toList(),
          )
        : QualitySelectorProxyApi.fromOrderedList2(
            qualities: qualities.map((e) => e.api).toList(),
            fallbackStrategy: fallbackStrategy.api,
          );
    return QualitySelectorImpl.internal(api);
  }
}

extension QualitySelectorX on QualitySelector {
  QualitySelectorProxyApi get api {
    final impl = this;
    if (impl is! QualitySelectorImpl) throw TypeError();
    return impl.api;
  }
}

extension QualitySelectorProxyApiX on QualitySelectorProxyApi {
  QualitySelector get impl => QualitySelectorImpl.internal(this);
}

File: camerax_android/android/src/main/kotlin/dev/zeekr/camerax_android/video/QualitySelectorImpl.kt
package dev.zeekr.camerax_android.video

import android.util.Size
import androidx.camera.core.CameraInfo
import androidx.camera.video.FallbackStrategy
import androidx.camera.video.Quality
import androidx.camera.video.QualitySelector
import dev.zeekr.camerax_android.CameraXApiPigeonProxyApiRegistrar
import dev.zeekr.camerax_android.PigeonApiQualitySelectorProxyApi

class QualitySelectorImpl(registrar: CameraXApiPigeonProxyApiRegistrar) : PigeonApiQualitySelectorProxyApi(registrar) {
    override fun from1(quality: Quality): QualitySelector {
        return QualitySelector.from(quality)
    }

    override fun from2(quality: Quality, fallbackStrategy: FallbackStrategy): QualitySelector {
        return QualitySelector.from(quality, fallbackStrategy)
    }

    override fun fromOrderedList1(qualities: List<Quality>): QualitySelector {
        return QualitySelector.fromOrderedList(qualities)
    }

    override fun fromOrderedList2(qualities: List<Quality>, fallbackStrategy: FallbackStrategy): QualitySelector {
        return QualitySelector.fromOrderedList(qualities, fallbackStrategy)
    }

    override fun getResolution(cameraInfo: CameraInfo, quality: Quality): Size? {
        return QualitySelector.getResolution(cameraInfo, quality)
    }
}